### ISSUE TITLE
Placeholder (or similar concept) with generic matching

### ISSUE THREAD

**Issue by cneijenhuis** [agourlay/cornichon#98]
Sometimes, I don't know the exact value my API would return, but I know that it has to return a value, and maybe I know that the value has to have a specific shape (e.g. an UUID or a Date).

Currently, I can use `body.path("id").isPresent` to assert that a value is there at all. This is somewhat fine, but doesn't have a nice reading flow, e.g.:

```
And assert body.whiteListing.is(
"""
{
  "firstName": "Foo",
  "lastName": "Bar",
  "isAwesome": true
}
""")
And assert body.path("id").isPresent
```

The reading flow is better if the `"id"` presence is checked within the body-assertion.

```
And assert body.whiteListing.is(
"""
{
  "id": <is-present>,
  "firstName": "Foo",
  "lastName": "Bar",
  "isAwesome": true
}
""")
```

It would be even better if we can define matchers for the shape of the fields, e.g.:

```
And assert body.whiteListing.is(
"""
{
  "id": "<any-uuid>",
  "firstName": "Foo",
  "lastName": "Bar",
  "isAwesome": <any-boolean>,
  "createdAt": "<any-iso-date>"
}
""")
```

I used the `<>` syntax from placeholders in the examples. It would also be completely valid to introduce another syntax for it, as it's a different concept (e.g. `[]`).

**Comment by agourlay** [comment]
Looks really good, thank you for the input!
I can definitely see how this could improve the user experience.

However it does require a non-trivial modification of the current model so no promises 😄

**Comment by peter-gerhard** [comment]
for primitive types it would be really cool but how should the matcher look for an object field?
would it be better to have something like `resolvesTo` which tries to instantiate a case class from it?

**Comment by agourlay** [comment]
Quick idea, as a first approach I would try require a Circe Decoder of the target type.
This will receives the captured text/json and try to instantiate an object.
It does not explode then it can be declared valid.

The lib would have a built-in set of matchers for common types and allow extension through a registration process as it is done for customer resolver.

**Comment by agourlay** [comment]
the issue is tackled here https://github.com/agourlay/cornichon/tree/body-string-matchers

**Comment by agourlay** [comment]
The last remaining issue is the usage of matchers without quotes such as ```any-int``` and ```any-boolean```.

The current solution uses quotes for all matchers which might be a bit misleading,

**Comment by agourlay** [comment]
@peter-gerhard I think I have changed my mind regarding the assertion function.
From 
```String => Boolean```
to
```Json => Boolean```

It will avoid to always reparse the input and will handle properly a bunch of edge cases such as ```"123"``` being validated as ```any-int```.

WDYT?

**Comment by peter-gerhard** [comment]
Cool stuff :)
I took a quick look at the reworks, looks good.
One thing I saw, I would expect '0' to match <<any-positive-integer>>, right?

**Comment by agourlay** [comment]
@peter-gerhard I don't know about zero, it does not belong to the positive nor negative set.

```
An integer is a whole number that can be either greater than 0, called positive, or less than 0, called negative. Zero is neither positive nor negative. Two integers that are the same distance from the origin in opposite directions are called opposites.
```

It may depends on your sources :bowtie:

**Comment by peter-gerhard** [comment]
@agourlay true that. idk i find it useful and consistent with `<random-positive-integer>`, but i can always define my custom `<<any-non-negative-integer>>` ;)

**Comment by agourlay** [comment]
I have done the predicate change, the code is much cleaner.
https://github.com/agourlay/cornichon/commit/9c9733e5f7d58e2576b1f3d607ec81b0a02e4812

I also added a description field to help our dear users :bowtie:

**Comment by agourlay** [comment]
fixed in 0.11
